### PR TITLE
Add ability to enable the MSVC address sanitizer via cmake

### LINKED ISSUE #29568: Enable MSVC address sanitizer if USE_SANITIZER set to Address
Microsoft have released an address sanitizer for Visual Studio in versions 16.7 onwards. Based on some initial tests this appears to work providing some care is taken around the compile flags. This issue is to enhance the Mantid cmake logic to set the required options in Visual Studio to run this new address sanitizer if the cmake option USE_SANITIZER=Address

image

I haven't so far been able to set the highlighted Yes\No option directly via cmake. But does look like I can set the compiler\linker options that result from this being set to Yes.

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

Microsoft have recently added an address sanitizer that checks for run time memory errors to VS 2019. 

https://devblogs.microsoft.com/cppblog/addresssanitizer-asan-for-windows-with-msvc/
https://devblogs.microsoft.com/cppblog/asan-for-windows-x64-and-debug-build-support/

These changes allow this to be configured in all of the Mantid targets using the existing USE_SANITIZER switch that was previously only relevant to the gcc compiler.

Excluded the gmock target from having the sanitizer set up applied because the gmock cmake set up specifies a W4 logging level that doesn't play nicely with the sanitizer. Since this isn't our code have assumed we don't need to sanitize it

**To test:**

1. Check you are running Visual Studio 2019 v16.7 or later
2. Run CMake with USE_SANITIZER set to Address. If using the CMake gui you need to press Configure, wait to complete, then press Generate
3. This should update most of the targets (ie Visual Studio projects) with some new settings
4. In Visual Studio go to Debug, Windows, Exception Settings and uncheck the exception with description "0xc0000005 0xC0000005" under the "Win32 Exceptions" heading
5. Add this (or similar) to your path environment variable:
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x64
6. Paste the following lines of code into one of the C++ units within the Framework (eg in the exec method of one of the Algorithms). These lines create an address issue that the sanitizer should find:
```
int* array = new int[100];
array[100] = 1;
```
7. Build the Framework and the corresponding unit test (eg in AlgorithmsTest) in RelWithDebInfo mode
8. Run the unit test from Visual Studio and check the debugger stops\finds this problem
9. Check the dev docs changes to the "Running Sanitizers" page adequately describe the set up. They cover the gcc set up and now the Visual Studio use of sanitizers

Fixes #29568. 

*This does not require release notes* because **the sanitizer is not active in release builds of Mantid**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: buildconfig/CMake/Sanitizers.cmake [hunk 1]
```diff
@@ -9,54 +9,72 @@ set_property(CACHE USE_SANITIZER PROPERTY STRINGS
...
+            add_compile_options($<$<OR:$<CONFIG:Debug>,$<CONFIG:RelWithDebInfo>>:/fsanitize=address>)
+            add_link_options("$<IF:$<CONFIG:Debug>,/wholearchive:clang_rt.asan_dbg_dynamic-x86_64.lib;/wholearchive:clang_rt.asan_dbg_dynamic_runtime_thunk-x86_64.lib;/wholearchive:vcasand.lib,/wholearchive:clang_rt.asan_dynamic-x86_64.lib;/wholearchive:clang_rt.asan_dynamic_runtime_thunk-x86_64.lib;/wholearchive:vcasan.lib>")
+            # Re additional switches enabled below on gcc:
+            # Equivalent of -fno-omit-frame-pointer in MSVC is /Oy- but documentation says it's not available for x64
+            # Don't think there is a VS equivalent of -fno-optimize-sibling-calls
+        else()
+            message(FATAL_ERROR "MSVC only supports address sanitizer")
+        endif()
+    else()
+        # Check and warn if we are not in a mode without debug symbols)
```
Comment (DavidFair) [inline buildconfig/CMake/Sanitizers.cmake:41]: Could we hoist this warning out of the if block, as it applies to both

File: buildconfig/CMake/Sanitizers.cmake [hunk 2]
```diff
@@ -9,54 +9,72 @@ set_property(CACHE USE_SANITIZER PROPERTY STRINGS
...
+        string(REPLACE "/O2" "/O1" CMAKE_C_FLAGS_RELWITHDEBINFO ${CMAKE_C_FLAGS_RELWITHDEBINFO})
+        set(CMAKE_CXX_FLAGS_RELWITHDEBINFO ${CMAKE_CXX_FLAGS_RELWITHDEBINFO} CACHE STRING "" FORCE)
+        set(CMAKE_C_FLAGS_RELWITHDEBINFO ${CMAKE_C_FLAGS_RELWITHDEBINFO} CACHE STRING "" FORCE)
 
-    elseif(${build_type_lower} MATCHES "relwithdebinfo")
+        if(${USE_SANITIZERS_LOWER} MATCHES "address")
+            add_compile_options($<$<OR:$<CONFIG:Debug>,$<CONFIG:RelWithDebInfo>>:/fsanitize=address>)
+            add_link_options("$<IF:$<CONFIG:Debug>,/wholearchive:clang_rt.asan_dbg_dynamic-x86_64.lib;/wholearchive:clang_rt.asan_dbg_dynamic_runtime_thunk-x86_64.lib;/wholearchive:vcasand.lib,/wholearchive:clang_rt.asan_dynamic-x86_64.lib;/wholearchive:clang_rt.asan_dynamic_runtime_thunk-x86_64.lib;/wholearchive:vcasan.lib>")
+            # Re additional switches enabled below on gcc:
+            # Equivalent of -fno-omit-frame-pointer in MSVC is /Oy- but documentation says it's not available for x64
```
Comment (DavidFair) [inline buildconfig/CMake/Sanitizers.cmake:35]: I don't think we need these 3 lines of comments. The optimisers both use are completely different so I would be more surprised if we had an equivalent `-fno-omit-frame-pointer` etc.

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor changes, otherwise looks good

**Comment by DavidFair** [comment]
I'm also getting:
LINK : fatal error LNK1356: cannot find library `clang_rt.asan_dbg_dynamic-x86_64.lib' specified to be whole archived

In debug mode, apparently this is not supported (which made me do a double take) so the docs need to say in RelWithDebMode. 
I wouldn't bother adding a CMake check as they plan to support it at some point.....

**Comment by DannyHindson** [comment]
> I'm also getting:
> LINK : fatal error LNK1356: cannot find library `clang_rt.asan_dbg_dynamic-x86_64.lib' specified to be whole archived
> In debug mode, apparently this is not supported (which made me do a double take) so the docs need to say in RelWithDebMode.
> I wouldn't bother adding a CMake check as they plan to support it at some point.....

It works for me in debug mode. Do you have that .lib file installed? For me it's in:

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\lib\x64

...and I think the linker finds it because of the change to the PATH env variable I describe in the testing steps

File: buildconfig/CMake/Sanitizers.cmake (hunk 1 above)
Comment (DannyHindson) [reply buildconfig/CMake/Sanitizers.cmake:41]: Sorry only just getting round to making these changes. I think this warning is only relevant to a single configuration generator such as ninja where you specify the build type? If you're using a multi configuration generator such as the Visual Studio generator then the configuration for all of the build types (release, debug etc) is created at the same time when you run cmake

**Review by DavidFair — APPROVED** [review]
Sorry about the delay in testing this again - it now works, I suspect a restart was required on the day I tested as I had updated VS version

This is a really useful tool to have so thanks :+1:

**Pull request merged**